### DISCUSSION TITLE
Variance and Std Deviation of a collection of Quantities..

### DISCUSSION THREAD

**Discussion by benhutchison** [erikerlandson/coulomb#447] [General]
Erik, have you ever needed to compute the variance and standard deviation of some `Quantity[V, U]`s?

The variance ends up in units `U * U`. I'm not sure if that's useful. Theoretically, taking the square root will mean the std dev has the right units, but there isn't a sqr root function ATM that will cleanly turn `U * U` into `U`, AFAICT?

My inclination, for now, is to coerce variance to type `U`.

**Comment by erikerlandson** [comment]
It's a good question - If I had collections for Quantity, then I'd probably (under the hood) compute stdv in unitless raw value types, but return the final output as `raw_stdv.withUnit[U]`, and similarly for variance `raw_var.withUnit[U^2]`

**Reply by erikerlandson** [reply]
Or just implement variance, and stdv is `(variance).pow[1/2]` at the end, which we can already do for quantity

**Reply by benhutchison** [reply]
Actually, this didnt seem to work out in practice. Finding square root of 4 isn't a clean 2 when using `pow[1/2]`.
 
```
scala> class Book
// defined class Book

scala> val bq = BigDecimal(2).withUnit[Book]
val bq: coulomb.Quantity[BigDecimal, Book] = 2

scala> val variance = (bq * bq)
val variance: coulomb.Quantity[BigDecimal, Book ^ 2] = 4

scala> variance.pow[1/2]
val res0: coulomb.Quantity[BigDecimal, Book] = 1.999999999999999999999999999999999
```

**Reply by erikerlandson** [reply]
That is interesting - I don't know a lot about BigDecimal representation, or for that matter the details of what the operations are doing here.  IIRC it does the right thing for Double.

**Reply by benhutchison** [reply]
It's not purely a BigDecimal thing, but I haven't yet had a chance to look into how `pow[1/2]` works under the hood. Anyway, easily worked around with a custom `qSqrt` extension.
```
scala> val bd = java.math.BigDecimal(2)
val bd: java.math.BigDecimal = 2

scala> (bd.multiply(bd)).sqrt(BigDecimal.defaultMathContext)
val res1: java.math.BigDecimal = 2
```

**Reply by erikerlandson** [reply]
The thing that's doing the square root should be this:
https://github.com/erikerlandson/coulomb/blob/scala3/spire/src/main/scala/coulomb/ops/algebra/spire/bigdecimal.scala#L27

**Reply by erikerlandson** [reply]
My recollection is, that I was assuming if people wanted roots, they were ok with floating point jitter, and I was trying to leverage the `fpow` "standard", which you can see forces the exponent from rational to fp, which is likely going to work against exact answers to rational powers.

We might reconsider this, it would require a redesign of the algebras I used to define powers.

**Reply by erikerlandson** [reply]
or we could do something for `q.sqrt` and maybe specialized algebras for that particular case

**Reply by erikerlandson** [reply]
I'm still slightly puzzled by why this is happening.  `1/2` has an exact fp representation, even if other fractions don't.

**Reply by benhutchison** [reply]
"I was assuming if people wanted roots, they were ok with floating point jitter" <- on consideration I think this is a decent first assumption. While Im working with financial data where to-the-cent accuracy is needed, the only use of square root shows up in stats over the transactions amounts, ie stddev.

The only reason I noticed this was a unit test where there was a contrived sqrt(4) that failed to eql 2. If I use epsilon equality like I probably ought to have been, the problem goes away.
